[rpc] Remove template on RRef and add Type to RRef creation

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30137 [jit] Add RRef to IValue and Jit
* **#30630 [rpc] Remove template on RRef and add Type to RRef creation**
* #30136 [jit] Enable IValue to hold a PyObject

Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19559729](https://our.internmc.facebook.com/intern/diff/D19559729)

wanchaol committed:
[rpc] Remvove template on RRef and add Type to RRef creation

Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 9732c67:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9732c67e720ebc296edced0cb7c05da1794da036).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3960858) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74868472))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
 
  git checkout -b <new-branch-name> 
 
HEAD is now at 9732c67e72 Update on "[rpc] Remvove template on RRef and add Type to RRef creation" 
+ git reset --hard 9732c67e720ebc296edced0cb7c05da1794da036 
HEAD is now at 9732c67e72 Update on "[rpc] Remvove template on RRef and add Type to RRef creation" 
+ git merge --allow-unrelated-histories --no-edit --no-ff 409151e1bb90c66546901d372f59930f72242ee9 
Auto-merging torch/csrc/jit/script/script_type_parser.cpp 
CONFLICT (content): Merge conflict in torch/csrc/jit/script/script_type_parser.cpp 
Auto-merging torch/csrc/distributed/rpc/py_rref.cpp 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -209,25 +207,32 @@ class RRef {
+
+  // type field to denote the type of the element that the RRef is holding
+  // it could be any TypePtr that JIT support, including PyObjectType
+  const TypePtr type_;
```
Was wondering whether TypeKind would be sufficient here? Is there a reason we use TypePtr over TypeKind?

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -47,26 +47,21 @@ class RRefContext {
   // one.
-  template <typename T>
-  std::shared_ptr<OwnerRRef<T>> getOrCreateOwnerRRef(const RRefId& rrefId);
+  std::shared_ptr<OwnerRRef> getOrCreateOwnerRRef(const RRefId& rrefId, c10::optional<TypePtr> opt_type={});
```
What are the cases where we don't need to pass in the type?

wanchaol replied:
Basically when we need to purely just get the RRef instead of potentially creating it, we don't need to pass in the type, otherwise we will need to pass the type in.

pritamdamania87 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -52,10 +40,10 @@ py::object PyRRef::toHere() {
-          std::static_pointer_cast<UserRRef<IValue>>(rref_)->toHere();
+          std::static_pointer_cast<UserRRef>(rref_)->toHere();
 
       {
```
Do we have unit tests that cover both of these cases?

wanchaol replied:
hmm i think we have the test cases already for pyObj and IValue cases.

wanchaol commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -78,38 +82,40 @@ RRefForkData RRefForkData::fromPyTuple(const py::tuple& t) {
+
   worker_id_t parent = t[PARENT_IDX].cast<worker_id_t>();
-  return RRefForkData(ownerId, rrefId, forkId, parent);
+  const std::string& typeStr = t[TYPE_IDX].cast<std::string>();
```
The problem of just using the TypeKind is that we could not distinguish class types from it. for the class types registered in the compilation unit, we could not know the concrete type information except TypeKind::classType. So I make this with string in the python pickling. There're also some concerns even on type str, i.e. two worker created two module class instances of the same `nn.Module`, but they might have different mangled type names, I am collecting those cases to ensure we have consistent serde for corner cases.

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

mrshenli reviewed (approved):
This LGTM! Code is more laconic. I added some minor comments inline. And lint failures are real, please address them too.

We probably can further make `owners_` to use `OwnerRRef`s and `pendingUsers_` to use `UserRRef`s. They weren't so before because those two classes are templatized. With this PR, we can make them more specific too. This does not need to be added into this PR. I can work on those when this is landed.  

Thanks @wanchaol !

mrshenli commented:
Type in Title "Remvove" -> "Remove"

mrshenli reviewed (changes requested):
As we discussed offline, changes will be made to generalize TypePtr serde. Besides, #31893 will likely land before this PR and might cause conflicts.

wanchaol commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -151,87 +133,64 @@ void RRefContext::delUser(
-    const RRefId& rrefId) {
+std::shared_ptr<OwnerRRef> RRefContext::getOrCreateOwnerRRef(
+    const RRefId& rrefId,
+    c10::optional<TypePtr> type) {
```
changed the API to force passing TypePtr as well.

wanchaol committed:
Update on "[rpc] Remvove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -25,6 +47,8 @@ PythonRpcHandler::PythonRpcHandler() {
   pyLoadReturnValue_ = getFunction(module, "_load_return_value");
   pySerialize_ = getFunction(module, "serialize");
   pyHandleException_ = getFunction(module, "_handle_exception");
+  typeParser_ = std::make_shared<jit::script::ScriptTypeParser>(
```
does this require GIL?

wanchaol replied:
probably require GIL when get the python_cu via `torch::jit::get_python_cu()`

mrshenli reviewed (approved):
Thanks for addressing comment. The latest change LGTM. clang-tidy lint failures are real though. 

Please coordinate with #31893 on the merge order.

wanchaol committed:
fix clang-tidy on "[rpc] Remove template on RRef and add Type to RRef creation"


Summary:

This remove template and all the specializations it have in rpc, we
universally use IValue as the inner value since we support making python
object to be hold inside IValue.

This will also ensure that we have the correct type information when
creating the RRef, we use the return type from the schema when creating
userRRef and OwnerRRef, it will enable IValue to always have the correct
type if the IValue is the RRef object (next PR)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19502235](https://our.internmc.facebook.com/intern/diff/D19502235)

[ghstack-poisoned]

zhaojuanmao commented:
failures are not relevant
